File: src/db/seed-docs.py
#!/usr/bin/env python3
"""
Insert 20 seed documents for Alex Johnson directly via psql.
Run: python3 src/db/seed-docs.py
"""
import hashlib
import json
import os
import subprocess
import sys

DB_URL = os.environ.get('DATABASE_URL', 'postgresql://sc_dev:dev_password@localhost:5438/studentcontext')
SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
JSON_PATH = os.path.join(SCRIPT_DIR, 'seed-documents.json')


def esc(s):
    return s.replace("'", "''")


def split_into_chunks(text, chunk_size=500, overlap=50):
    trimmed = text.strip()
    if not trimmed:
        return []
    est_tokens = len(trimmed) // 4 + 1
    if est_tokens <= chunk_size:
        return [{'text': trimmed, 'tokenCount': est_tokens}]

    chunk_chars = chunk_size * 4
    overlap_chars = overlap * 4
    chunks = []
    start = 0
    while start < len(trimmed):
        end = min(start + chunk_chars, len(trimmed))
        if end < len(trimmed):
            search_start = max(end - 200, start + 1)
            last_space = trimmed.rfind(' ', search_start, end)
            last_newline = trimmed.rfind('\n', search_start, end)
            bp = max(last_space, last_newline)
            if bp > search_start:
                end = bp + 1
        ct = trimmed[start:end].strip()
        if ct:
            chunks.append({'text': ct, 'tokenCount': len(ct) // 4 + 1})
        if end >= len(trimmed):
            break
        new_start = end - overlap_chars
        if new_start <= start:
            new_start = end
        start = new_start
    return chunks


def main():
    with open(JSON_PATH) as f:
        docs = json.load(f)

    # Build SQL
    lines = []
    lines.append("BEGIN;")
    lines.append("")
    lines.append("DO $$")
    lines.append("DECLARE")
    lines.append("  v_sid UUID;")
    lines.append("  v_bid UUID;")
    lines.append("  v_did UUID;")
    lines.append("BEGIN")
    lines.append("  SELECT u.id, u.board_id INTO v_sid, v_bid")
    lines.append("  FROM users u WHERE u.email = 'alex.johnson@tvdsb.on.ca';")
    lines.append("  IF v_sid IS NULL THEN")
    lines.append("    RAISE EXCEPTION 'Alex Johnson not found. Run npm run db:seed first.';")
    lines.append("  END IF;")
    lines.append(f"  RAISE NOTICE 'Inserting {len(docs)} documents for student %', v_sid;")
    lines.append("")

    for doc in docs:
        content_hash = hashlib.sha256(doc['content'].encode()).hexdigest()
        chunks = split_into_chunks(doc['content'])
        meta_json = json.dumps(doc['metadata'])

        lines.append(f"  -- {doc['title']}")
        lines.append(f"  IF NOT EXISTS (SELECT 1 FROM documents WHERE hash = '{content_hash}') THEN")
        lines.append(f"    INSERT INTO documents (student_id, board_id, source, title, content, academic_year, sensitivity, metadata, hash)")
        lines.append(f"    VALUES (v_sid, v_bid, '{doc['source']}', '{esc(doc['title'])}', '{esc(doc['content'])}', '2025-2026', '{doc['sensitivity']}', '{esc(meta_json)}', '{content_hash}')")
        lines.append(f"    RETURNING id INTO v_did;")

        for i, chunk in enumerate(chunks):
            lines.append(f"    INSERT INTO chunks (document_id, student_id, board_id, content, chunk_index, token_count, sensitivity, metadata)")
            lines.append(f"    VALUES (v_did, v_sid, v_bid, '{esc(chunk['text'])}', {i}, {chunk['tokenCount']}, '{doc['sensitivity']}', '{{}}');")

        lines.append(f"    RAISE NOTICE '  [ok] {esc(doc['title'])} — {len(chunks)} chunks';")
        lines.append(f"  ELSE")
        lines.append(f"    RAISE NOTICE '  [skip] {esc(doc['title'])} (duplicate)';")
        lines.append(f"  END IF;")
        lines.append("")

    lines.append("END $$;")
    lines.append("")
    lines.append("COMMIT;")

    sql = '\n'.join(lines)

    # Run via psql
    result = subprocess.run(
        ['psql', DB_URL, '-v', 'ON_ERROR_STOP=1'],
        input=sql, capture_output=True, text=True
    )

    # Print notices (document status)
    for line in result.stderr.split('\n'):
        if 'NOTICE' in line:
            msg = line.split('NOTICE:')[-1].strip() if 'NOTICE:' in line else line
            print(msg)

    if result.returncode != 0:
        err = result.stderr
        if 'ERROR' in err:
            for line in err.split('\n'):
                if 'ERROR' in line:
                    print(f"Error: {line}", file=sys.stderr)
        sys.exit(1)

    print("\nDone.")


if __name__ == '__main__':
    main()

File: src/db/seed-docs.mjs
#!/usr/bin/env node
/**
 * Standalone document seeder — inserts 20 documents + chunks for Alex Johnson.
 * Uses pure node:pg to avoid tsx memory overhead.
 * Run: node src/db/seed-docs.mjs
 */
import crypto from 'node:crypto';
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import pg from 'pg';

const __dirname = path.dirname(fileURLToPath(import.meta.url));

const DATABASE_URL = process.env.DATABASE_URL || 'postgresql://sc_dev:dev_password@localhost:5438/studentcontext';

const pool = new pg.Pool({ connectionString: DATABASE_URL });

// Minimal chunker (same logic as src/ingestion/chunker.ts)
function splitIntoChunks(text, chunkSize = 500, overlap = 50) {
  if (!text || text.trim().length === 0) return [];
  const trimmed = text.trim();
  const estimatedTokens = Math.ceil(trimmed.length / 4);
  if (estimatedTokens <= chunkSize) {
    return [{ text: trimmed, tokenCount: estimatedTokens }];
  }
  const chunkChars = chunkSize * 4;
  const overlapChars = overlap * 4;
  const chunks = [];
  let start = 0;
  while (start < trimmed.length) {
    let end = Math.min(start + chunkChars, trimmed.length);
    if (end < trimmed.length) {
      const searchStart = Math.max(end - 200, start + 1);
      const lastSpace = trimmed.lastIndexOf(' ', end);
      const lastNewline = trimmed.lastIndexOf('\n', end);
      const breakPoint = Math.max(lastSpace, lastNewline);
      if (breakPoint > searchStart) end = breakPoint + 1;
    }
    const chunkText = trimmed.slice(start, end).trim();
    if (chunkText.length > 0) {
      chunks.push({ text: chunkText, tokenCount: Math.ceil(chunkText.length / 4) });
    }
    if (end >= trimmed.length) break;
    start = end - overlapChars;
    if (start <= (chunks.length > 0 ? end - chunkChars : 0)) start = end;
  }
  return chunks;
}

async function main() {
  // Get Alex's ID and board ID
  const userResult = await pool.query(
    `SELECT u.id AS student_id, u.board_id FROM users u WHERE u.email = 'alex.johnson@tvdsb.on.ca'`
  );
  if (userResult.rows.length === 0) {
    console.error('Error: Alex Johnson not found. Run `npm run db:seed` first (it creates users).');
    process.exit(1);
  }
  const { student_id: studentId, board_id: boardId } = userResult.rows[0];
  console.log(`Student: Alex Johnson (${studentId})`);

  // Load documents from JSON
  const jsonPath = path.join(__dirname, 'seed-documents.json');
  const docs = JSON.parse(fs.readFileSync(jsonPath, 'utf8'));
  console.log(`\nIngesting ${docs.length} documents...`);

  const client = await pool.connect();
  try {
    await client.query('BEGIN');

    let totalChunks = 0;
    let duplicates = 0;

    for (const doc of docs) {
      const contentHash = crypto.createHash('sha256').update(doc.content).digest('hex');

      // Check duplicate
      const existing = await client.query('SELECT id FROM documents WHERE hash = $1', [contentHash]);
      if (existing.rows.length > 0) {
        duplicates++;
        console.log(`  [skip] ${doc.title} (duplicate)`);
        continue;
      }

      // Insert document
      const docResult = await client.query(
        `INSERT INTO documents (student_id, board_id, source, title, content, academic_year, sensitivity, metadata, hash)
         VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9)
         RETURNING id`,
        [studentId, boardId, doc.source, doc.title, doc.content, '2025-2026', doc.sensitivity, JSON.stringify(doc.metadata), contentHash]
      );
      const docId = docResult.rows[0].id;

      // Chunk and insert
      const chunks = splitIntoChunks(doc.content);
      for (let i = 0; i < chunks.length; i++) {
        await client.query(
          `INSERT INTO chunks (document_id, student_id, board_id, content, chunk_index, token_count, sensitivity, metadata)
           VALUES ($1, $2, $3, $4, $5, $6, $7, $8)`,
          [docId, studentId, boardId, chunks[i].text, i, chunks[i].tokenCount, doc.sensitivity, '{}']
        );
        totalChunks++;
      }

      console.log(`  [ok]   ${doc.title} — ${chunks.length} chunks`);
    }

    await client.query('COMMIT');
    console.log(`\nDone: ${docs.length - duplicates} documents, ${totalChunks} chunks`);
    if (duplicates > 0) console.log(`  (${duplicates} duplicates skipped)`);
  } catch (err) {
    await client.query('ROLLBACK');
    throw err;
  } finally {
    client.release();
  }

  await pool.end();
}

main().catch((err) => { console.error(err); process.exit(1); });
